### PR TITLE
Simplify the paths the MSI uses

### PR DESCRIPTION
**Pull request by TravisEz13**
## PR Summary

Simplify the paths the MSI uses
    - Implementation of PowerShell/PowerShell-RFC#115 (If anything changes in the RFC, we will treat it as a bug, and fix it later)
    - Update registry and directory paths to use `6` for the version for stable and `6-preview` for a preview release
    - Add checkbox to set path
        - default checkbox to off for preview builds and on for stable builds  

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: assets/Product.wxs [hunk 1]
```diff
@@ -9,26 +9,26 @@
   <!-- UpgradeCode GUID MUST REMAIN SAME THROUGHOUT ALL VERSIONS, otherwise, updates won't occur. -->
   <?if $(sys.BUILDARCH)=x64?>
     <?define UpgradeCode = "$(env.UpgradeCodeX64)" ?>
-    <?define ProductSemanticVersionWithNameAndOptionalArchitecture = "$(var.ProductName) $(env.ProductSemanticVersion)"?>
-    <?define ExplorerContextMenuDialogText = "&$(var.ProductName) $(env.ProductSemanticVersion)"?>
+    <?define ProductSimpleVersionWithNameAndOptionalArchitecture = "$(var.ProductName) $(env.SimpleProductVersion)"?>
```
Comment (adityapatwardhan) [inline assets/Product.wxs]: Consider adding architecture here.

File: test/packaging/windows/msi.tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+function Invoke-Msiexec {
+    param(
+        [Parameter(ParameterSetName = 'Install', Mandatory)]
+        [Switch]$Install,
+
+        [Parameter(ParameterSetName = 'Uninstall', Mandatory)]
+        [Switch]$Uninstall,
+
+        [Parameter(Mandatory)]
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Add a `Test-Path` in validate script

File: test/packaging/windows/msi.tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        foreach ($key in $Properties.Keys) {
+            $additionalOptions += "$key=$($Properties.$key)"
+        }
+    }
+
+    $argumentList = "$switch $MsiPath /quiet /l*vx $script:msiLog $additionalOptions"
+    $msiExecProcess = Start-Process msiexec.exe -Wait -ArgumentList $argumentList -NoNewWindow -PassThru
+    if ($msiExecProcess.ExitCode -ne 0) {
+        $exitCode = $msiExecProcess.ExitCode
+        throw "$action MSI failed and returned error code $exitCode. MSI Log was uploaded as artifact."
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Remove part about `MSI log uploaded`

File: test/packaging/windows/msi.tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+Describe -Name "Windows MSI" -Fixture {
+    BeforeAll {
+        $script:msiX64Path = $env:PsMsiX64Path
+        $script:beforePath = @(([System.Environment]::GetEnvironmentVariable('PATH', 'MACHINE')) -split ';' |
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Add a comment

File: test/packaging/windows/msi.tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                Where-Object {$_ -like '*files\powershell*'})
+
+        $resolvedTestDrive = (Resolve-Path "Testdrive:\").providerPath
+        $script:msiLog = Join-Path -Path $resolvedTestDrive -ChildPath 'msilog.txt'
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Maybe use `$TestDrive`

File: test/packaging/windows/msi.tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    $msiExecProcess = Start-Process msiexec.exe -Wait -ArgumentList $argumentList -NoNewWindow -PassThru
+    if ($msiExecProcess.ExitCode -ne 0) {
+        $exitCode = $msiExecProcess.ExitCode
+        throw "$action MSI failed and returned error code $exitCode. MSI Log was uploaded as artifact."
+    }
+}
+
+Describe -Name "Windows MSI" -Fixture {
+    BeforeAll {
+        $script:msiX64Path = $env:PsMsiX64Path
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Convert script variable to regular variable as definition in BeforeAll are guaranteed to be present throughout Describe.

File: test/packaging/windows/msi.tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            } else {
+                Copy-Item -Path $script:msiLog -Destination $env:temp -Force
+                Write-Verbose "MSI log is at $env:temp\msilog.txt" -Verbose
+            }
+            $script:uploadedLog = $true
+        }
+    }
+
+    Context "Add Path disabled" {
+        It "MSI should install without error" -Skip:(!(Test-Elevated) -or !$script:msiX64Path) {
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Remove check for existence of `msiX64Path`

File: test/packaging/windows/msi.tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                Invoke-MsiExec -Install -MsiPath $msiX64Path -Properties @{ADD_PATH = 0}
+            } | Should -Not -Throw
+        }
+
+        It -name "MSI should have not be updated path" -Skip:(!(Test-Elevated) -or !$script:msiX64Path) -test {
```
Comment (adityapatwardhan) [inline test/packaging/windows/msi.tests.ps1]: Remove `-name` also `-test`

File: tools/appveyor.psm1 [hunk 9]
```diff
@@ -486,36 +486,20 @@ function Invoke-AppveyorFinish
...
-        $msiLog = Join-Path (Get-Location) 'msilog.txt'
-        $msiExecProcess = Start-Process msiexec.exe -Wait -ArgumentList "/I $msi /quiet /l*vx $msiLog" -NoNewWindow -PassThru
-        if ($msiExecProcess.ExitCode -ne 0)
-        {
-            Push-AppveyorArtifact msiLog.txt
-            $exitCode = $msiExecProcess.ExitCode
-            throw "MSI installer failed and returned error code $exitCode. MSI Log was uploaded as artifact."
-        }
-        Write-Log "MSI smoke test was successful"
+        # Install the lastest Pester and import it
```
Comment (adityapatwardhan) [inline tools/appveyor.psm1]: lastest -> latest

File: tools/appveyor.psm1 [hunk 10]
```diff
@@ -486,36 +486,20 @@ function Invoke-AppveyorFinish
...
-            Where-Object {$_ -like '*files\powershell*' -and $_ -notin $beforePath}
+        # start the packaging tests and get the results
+        $packagingTestResult = Invoke-Pester -Script (Join-Path $repoRoot '.\test\packaging\windows\') -PassThru
 
-        if(!$psPath)
+        # fail the CI job if the tests failed, or nothing passed
+        if($packagingTestResult.FailedCount -ne 0 -or !$packagingTestResult.PassedCount)
         {
-            throw "MSI did not add powershell to path"
+            throw "Packaging tests failed"
```
Comment (adityapatwardhan) [inline tools/appveyor.psm1]: Throw number of tests failed / passed.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: assets/Product.wxs (hunk 1 above)
Comment (TravisEz13) [reply assets/Product.wxs]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 2 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 3 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 4 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 5 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 6 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 7 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: test/packaging/windows/msi.tests.ps1 (hunk 8 above)
Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: fixed

File: tools/appveyor.psm1 (hunk 9 above)
Comment (TravisEz13) [reply tools/appveyor.psm1]: fixed

File: tools/appveyor.psm1 (hunk 10 above)
Comment (TravisEz13) [reply tools/appveyor.psm1]: fixed

File: test/packaging/windows/msi.tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        $beforePath = @(([System.Environment]::GetEnvironmentVariable('PATH', 'MACHINE')) -split ';' |
+                Where-Object {$_ -like '*files\powershell*'})
+
+        $msiLog = Join-Path -Path $TestDrive -ChildPath 'msilog.txt'
+
+        foreach ($pathPart in $beforePath) {
+            Write-Warning "Found existing PowerShell path: $pathPart"
+        }
+
+        if (!(Test-Elevated)) {
```
Comment (daxian-dbw) [inline test/packaging/windows/msi.tests.ps1]: Can't we depend on the `RequireAdminitstrator` tag?

Comment (TravisEz13) [reply test/packaging/windows/msi.tests.ps1]: No, I would have to build the infrastructure for that.  It depends on `Start-PSPester` and this isn't like the product tests.  At least for the MSI tests, everything will require admin.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by joeyaiello** [comment]
Per offline conversation with @TravisEz13, we should merge this as experimental into the working 6.1 branch, and any discrepancies with the eventually accepted RFC should get fixed as bugs.

We should also gate backporting this fix to 6.0 on acceptance of that RFC given that this is experimental behavior.

**Pull request merged**